From 2ed2bdf784a09ba043c315705f1450ee55a729e6 Mon Sep 17 00:00:00 2001
From: Avinash Kumar <avikumar@habana.ai>
Date: Mon, 22 Nov 2021 11:52:59 +0200
Subject: [PATCH] [SW-62043] greco: code reloc, sram bin creation

Relocation of some of greco mgmt drivers
from iccm to sram needed to save iccm. Using
objcpy relocated code is copied to -sram.bin.

Signed-off-by: Avinash Kumar <avikumar@habana.ai>
---

diff --git a/board/habanalabs/gaudi2/post-image-fit.sh b/board/habanalabs/gaudi2/post-image-fit.sh
index 5ac3aa0..164d470 100644
--- a/board/habanalabs/gaudi2/post-image-fit.sh
+++ b/board/habanalabs/gaudi2/post-image-fit.sh
@@ -55,6 +55,7 @@
 	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-pid-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-dccm.bin.le
 	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-iccm.bin.le
 	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-dccm.bin.le
+	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-sram.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-sram.bin.le
 
 	return
 }
diff --git a/board/habanalabs/gaudi2_arc/post-image-spi.sh b/board/habanalabs/gaudi2_arc/post-image-spi.sh
index 6049f6d..edb3ff3 100755
--- a/board/habanalabs/gaudi2_arc/post-image-spi.sh
+++ b/board/habanalabs/gaudi2_arc/post-image-spi.sh
@@ -44,6 +44,10 @@
 	# Align binary to 4 bytes
 	dd status=none conv=sync bs=4 if=${BINARIES_DIR}/zephyr-${PREFIX}-dccm.bin > \
 	${BINARIES_DIR}/${DEV_PREFIX}-zephyr-${PREFIX}-dccm.bin
+
+	/* Create dummy 4 bytes file for gaudi2 sram bin */
+	dd status=none conv=sync bs=4 if=/dev/zero count=1 > \
+	${BINARIES_DIR}/${DEV_PREFIX}-zephyr-${PREFIX}-sram.bin
 }
 
 function hl_pldm_hex {
diff --git a/board/habanalabs/greco/post-image-fit.sh b/board/habanalabs/greco/post-image-fit.sh
index d3fc4c7..d5d4b34 100755
--- a/board/habanalabs/greco/post-image-fit.sh
+++ b/board/habanalabs/greco/post-image-fit.sh
@@ -51,6 +51,7 @@
 	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-pid-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-dccm.bin.le
 	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-iccm.bin.le
 	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-dccm.bin.le
+	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-sram.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-sram.bin.le
 
 	# copy img-pointer, ppboot and preboot binaries from other HL targets required for boot fit creation
 	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_NAME}-ppboot-pcie_phy_fw.bin.le ${FILEPATH}/${DEV_NAME}-ppboot.bin.le
diff --git a/board/habanalabs/greco_arc/post-image-spi.sh b/board/habanalabs/greco_arc/post-image-spi.sh
index daab00c..d0a800b 100755
--- a/board/habanalabs/greco_arc/post-image-spi.sh
+++ b/board/habanalabs/greco_arc/post-image-spi.sh
@@ -25,6 +25,15 @@
 	# Align binary to 4 bytes
 	dd status=none conv=sync bs=4 if=${BINARIES_DIR}/zephyr-${PREFIX}-dccm.bin > \
 	${BINARIES_DIR}/${DEV_PREFIX}-zephyr-${PREFIX}-dccm.bin
+
+	# create sram binary
+	${OBJCOPY} -S --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON \
+	--remove-section=.eh_frame --only-section=_SRAM_TEXT_SECTION_NAME --only-section=_SRAM_RODATA_SECTION_NAME \
+	${BINARIES_DIR}/zephyr_${PREFIX}.elf ${BINARIES_DIR}/zephyr-${PREFIX}-sram.bin
+
+	# Align binary to 4 bytes
+	dd status=none conv=sync bs=4 if=${BINARIES_DIR}/zephyr-${PREFIX}-sram.bin > \
+	${BINARIES_DIR}/${DEV_PREFIX}-zephyr-${PREFIX}-sram.bin
 }
 
 pushd ${BINARIES_DIR}
